include/envoy/stats/symbol_table.h

```diff
@@ -1,13 +1,157 @@
+ * somewhat similar to UTF-8, with a variable-length array of uint8_t. See the
+ * implementation for details.
+ *
+ * StatNameStorage can be used to manage memory for the byte-encoding. Not all
+ * StatNames are backed by StatNameStorage -- the storage may be inlined into
+ * another object such as HeapStatData. StaNameStorage is not fully RAII --
+ * instead the owner must call free(SymbolTable&) explicitly before
+ * StatNameStorage is destructed. This saves 8 bytes of storage per stat,
+ * relative to holding a SymbolTable& in each StatNameStorage object.
+ *
+ * A StatName is a copyable and assignable reference to this storage. It does
+ * not own the storage or keep it alive via reference counts; the owner must
+ * ensure the backing store lives as long as the StatName.
+ *
+ * The underlying Symbol / SymbolVec data structures are private to the
+ * impl. One side effect of the non-monotonically-increasing symbol counter is
+ * that if a string is encoded, the resulting stat is destroyed, and then that
+ * same string is re-encoded, it may or may not encode to the same underlying
+ * symbol.
+ */
+class SymbolTable {
+public:
+  virtual ~SymbolTable() = default;
+
+  /**
+   * Encodes a stat name using the symbol table, returning a SymbolEncoding. The
+   * SymbolEncoding is not intended for long-term storage, but is used to help
+   * allocate and StatName with the correct amount of storage.
+   *
+   * When a name is encoded, it bumps reference counts held in the table for
+   * each symbol. The caller is responsible for creating a StatName using this
+   * SymbolEncoding and ultimately disposing of it by calling
+   * StatName::free(). Otherwise the symbols will leak for the lifetime of the
+   * table, though they won't show up as a C++ leaks as the memory is still
+   * reachable from the SymolTable.
+   *
+   * @param name The name to encode.
+   * @return SymbolEncoding the encoded symbols.
+   */
+  virtual SymbolEncoding encode(absl::string_view name) PURE;
+
+  /**
+   * @return uint64_t the number of symbols in the symbol table.
+   */
+  virtual uint64_t numSymbols() const PURE;
+
+  /**
+   * Decodes a vector of symbols back into its period-delimited stat name. If
+   * decoding fails on any part of the symbol_vec, we release_assert and crash
+   * hard, since this should never happen, and we don't want to continue running
+   * with a corrupt stats set.
+   *
+   * @param stat_name the stat name.
+   * @return std::string stringifiied stat_name.
+   */
+  virtual std::string toString(const StatName& stat_name) const PURE;
+
+  /**
+   * Deterines whether one StatName lexically precedes another. Note that
+   * the lexical order may not exactly match the lexical order of the
+   * elaborated strings. For example, stat-name of "-.-" would lexically
+   * sort after "---" but when encoded as a StatName would come lexically
+   * earlier. In practice this is unlikely to matter as those are not
```

Comment:
Why do we have lexical comparison in the first place? Is this just to use in `std::map`, or should there be some documented ordering?